### ISSUE TITLE
Odd doubling of paths

### ISSUE THREAD

**Issue by Jamie-sas** [gogradle/gogradle#140]
So, sample project 'odd'.  Structure is

odd
 |__ build.gradle
 |__ src
 ____|__ odd
 ________|__ odd.go

Contents of odd.go is the example 'hello world' from the main golang.org page.

build.gradle file is just the first part of the 'Getting Started' page here, adjusted for this project:

-------------------------------------------------------------
plugins {
    id 'com.github.blindpirate.gogradle' version '0.7.0'
}

golang {
    packagePath = 'odd'
}
-------------------------------------------------------------

Using gradle 4.0.2.

I did 'gradle init' then 'gradle build', and it failed like this:

can't load package: package odd: no buildable Go source files in /odd/.gogradle/project_gopath/src/odd

Well, it _is_ telling the truth, but only because gogradle appears to have doubled the paths...

In other words, under /odd/.gogradle/project_gopath/src/odd, the source is actually in /odd/.gogradle/src/odd/src/odd/odd.go

So... why is it copying the files somewhere it can't find them to build them?

**Comment by blindpirate** [comment]
Well thank you for raising this, it's an excellent question.

In go community, nobody will structure code like your `odd` because they have a totally different convention from Maven (`src/main/java`). Generally speaking, They will put their go code directly in the project directory. 

Although Gradle recommend putting code in `src` directory, actually go developers won't do that. See [Github top 1000 repository](http://github-rank.com/star?language=Go) and [Gogradle examples](https://github.com/gogradle/examples) for more examples.

Back to your question. Actually you should structure code like this:

```
odd
|__ build.gradle
|__ odd.go
```

This should work.

Please let me know if there's anything I didn't explain clear enough.

**Comment by Jamie-sas** [comment]
While that may be correct for the projects on Github and your examples, it is incorrect for the code I am working with here.

Yes, your example is clear enough, and it does make the simple project I used as an example work.

It doesn't help me, though, since the code I am trying to get working is code that was written a while ago in a corporate environment, where I have no control of the structure.  I have to try to make this work with the structure as it is now, where people more closely (though still not exactly) followed the code layout guidelines documented on golang.org.

With that in mind, is there any way to get gogradle to work with the "standard" layout?

**Comment by blindpirate** [comment]
Oh, I understand.

In your scenario, you should add some code:

```
golang {
    packagePath = 'odd'
}
build {
    go 'build -o ./myOutput odd/src/odd'
}
```

See [here](https://github.com/gogradle/gogradle/blob/master/docs/tasks.md#build) for more details.

Please let me know if there's anything unclear.

**Comment by Jamie-sas** [comment]
That makes it work in the simple example.  I'll have to try it on the real project.

Would be nicer, of course, if it didn't require the extra "clunky" command, but it is a viable workaround.

**Comment by blindpirate** [comment]
I'm always try to make things elegant. So in your case, is there any other ways to make it elegant or neat? I mean, since your main package doesn't locate in project root, which kind of configuration can be more satisfactory for you? I'd like to hear your advice. 

It would be great if you can write a more "ideal" `build.gradle` to make it work. Thanks.

**Comment by Jamie-sas** [comment]
2 things immediately come to mind... if there _is_ a 'src' subdir, automatically default to the last example you gave.  Automatically handle that case for multiple platforms, as well, as you do for multiple platforms with the source in the root.

Those 2 changes alone would have gotten me much further along.  :-)

**Comment by blindpirate** [comment]
Well, in your case, actually `src` is not enough, we have to recognize `src/odd`, which may be a bit hard. Multi-platform are not handled automatically partially because for large project it would be extremely slow to compile and link for a specific platform.

**Comment by Jamie-sas** [comment]
What I meant was...

If I include (for example)

targetPlatform = ['linux-amd64', 'windows-amd64']

in my build.gradle for my small example, with the go code in the root, it will automatically handle running the multiple builds and naming the files appropriately.

Using the example you gave, with the code code in src/odd, and the same targetPlatform property, I had to create 'myOutput' as a directory, then change your example build line from

go 'build -o ./myOutput odd/src/odd'

to

go 'build -o ./myOutput/${GOOS}_${GOARCH}_${PROJECT_NAME}${GOEXE} odd/src/odd'

that was the 2nd thing I wanted handled automatically.

For handling src, couldn't you just get a recursive FileTree with a filter to only return .go files?  That would give the path needed.

**Comment by Jamie-sas** [comment]
An example fileTree call for a plugin I wrote, modified for use here;

def files = project.fileTree(dir: 'src', include: '**/*.go')

**Comment by blindpirate** [comment]
Thanks for the advice, I'll consider it. If there's no more questions, I'll close this. Please feel free to reopen it if any issue still exists.
